File: java/src/main/java/vsdk/toolkit/material/Xxdf.java
package vsdk.toolkit.material;

/**
General definitions for edf, brdf, btdf, etc.
*/

/**
Some general functions regarding edf, brdf, btdf, bsdf
*/

/**
Calculate the ideal reflected ray direction (independent of the brdf)
*/

/**
Calculate the perfect refracted ray direction.
Sets totalInternalReflection to true or false accordingly.
Cfr. [GLAS1989] An Introduction to Raytracing (Glassner)
*/

// Only real part of n for now

import vsdk.toolkit.common.linealAlgebra.Numeric;
import vsdk.toolkit.common.linealAlgebra.Vector3D;

public class Xxdf {
    public static final float PHONG_LOWEST_SPECULAR_EXP = 250.0f;

    public static Vector3D idealReflectedDirection(Vector3D in, Vector3D normal) {
        double tmp = 2.0 * normal.dotProduct(in);
        Vector3D result = new Vector3D();

        result.scaledCopy((float)tmp, normal);
        result.subtraction(in, result);
        result.normalize(Numeric.EPSILON_FLOAT);

        return result;
    }

    public static Vector3D idealRefractedDirection(
        Vector3D in,
        Vector3D normal,
        RefractionIndex inIndex,
        RefractionIndex outIndex,
        boolean[] totalInternalReflection) {
        float refractionIndex = inIndex.getNr() / outIndex.getNr();
        float ci = -in.dotProduct(normal);
        float ct2 = 1.0f + refractionIndex * refractionIndex * (ci * ci - 1.0f);

        if (ct2 < 0.0f) {
            if (totalInternalReflection != null && totalInternalReflection.length > 0) {
                totalInternalReflection[0] = true;
            }
            return idealReflectedDirection(in, normal);
        }

        if (totalInternalReflection != null && totalInternalReflection.length > 0) {
            totalInternalReflection[0] = false;
        }

        float ct = (float)Math.sqrt(ct2);
        float normalScale = refractionIndex * ci - ct;

        Vector3D result = new Vector3D();
        result.scaledCopy(refractionIndex, in);
        result.sumScaled(result, normalScale, normal);
        result.normalize(Numeric.EPSILON_FLOAT);

        return result;
    }
}

File: typescript/src/main/vsdk/toolkit/material/Xxdf.ts
import { Numeric } from "../common/linealAlgebra/Numeric";
import { Vector3D } from "../common/linealAlgebra/Vector3D";
import { RefractionIndex } from "./RefractionIndex";

export class Xxdf {
  public static readonly PHONG_LOWEST_SPECULAR_EXP = 250.0;

  public static idealReflectedDirection(inDirection: Vector3D, normal: Vector3D): Vector3D {
    const tmp = 2.0 * normal.dotProduct(inDirection);
    const result = new Vector3D();

    result.scaledCopy(tmp, normal);
    result.subtraction(inDirection, result);
    result.normalize(Numeric.EPSILON_FLOAT);

    return result;
  }

  public static idealRefractedDirection(
    inDirection: Vector3D,
    normal: Vector3D,
    inIndex: RefractionIndex,
    outIndex: RefractionIndex,
    totalInternalReflection: boolean[] | null
  ): Vector3D {
    const refractionIndex = inIndex.getNr() / outIndex.getNr();
    const ci = -inDirection.dotProduct(normal);
    const ct2 = 1.0 + refractionIndex * refractionIndex * (ci * ci - 1.0);

    if (ct2 < 0.0) {
      if (totalInternalReflection !== null && totalInternalReflection.length > 0) {
        totalInternalReflection[0] = true;
      }
      return Xxdf.idealReflectedDirection(inDirection, normal);
    }

    if (totalInternalReflection !== null && totalInternalReflection.length > 0) {
      totalInternalReflection[0] = false;
    }

    const ct = globalThis.Math.sqrt(ct2);
    const normalScale = refractionIndex * ci - ct;

    const result = new Vector3D();
    result.scaledCopy(refractionIndex, inDirection);
    result.sumScaled(result, normalScale, normal);
    result.normalize(Numeric.EPSILON_FLOAT);

    return result;
  }
}
